base/scratch_buffer.jl

```diff
@@ -0,0 +1,76 @@
+# NOTE: ryu/Ryu.jl: neededdigits(::Type{Float64}) = 309 + 17
+const SCRATCH_BUFFER_MIN_LENGTH = 309 + 17
+
+struct ScratchBuffer
+    p::Ptr{UInt8}
+    length::UInt64
+end
+
+# NOTE: similar to pcre.jl, Threads module isn't available at the time of loading
+_tid() = Int(ccall(:jl_threadid, Int16, ())) + 1
+_nth() = Int(unsafe_load(cglobal(:jl_n_threads, Cint)))
+
+const SCRATCH_BUFFERS = map(1:_nth()) do _
+    Vector{UInt8}(undef, SCRATCH_BUFFER_MIN_LENGTH), false
+end
+
+function init_scratch_buffers()
+    while length(SCRATCH_BUFFERS) < _nth()
+        push!(SCRATCH_BUFFERS, (Vector{UInt8}(undef, SCRATCH_BUFFER_MIN_LENGTH), false))
+    end
+    nothing
+end
+
+function with_scratch_buffer(f, n::Int)
+    buf, buf_inuse = @inbounds SCRATCH_BUFFERS[_tid()]
```

Comment:
The number of threads can now change, so we should check here instead of using `__init__`.